src/nvim/insexpand.c

```diff
@@ -1960,6 +1961,16 @@ static bool set_ctrl_x_mode(const int c)
   return retval;
 }
 
+// Restore extmarks, should be called if the original text is restored.
+static void ins_compl_undo_extmark(void)
+{
+  u_header_T *uhp = curbuf->b_u_curhead ? curbuf->b_u_curhead : curbuf->b_u_newhead;
```

Comment:
Unsure. TBH I'm not familiar with the undo code. This was my best guess after reading some of it.

Reply:
> :undojoin | normal ifoo<c-p>

Yeah this does seem to undo earlier extmark changes now.

Reply:
Yeah not sure if it is feasible to manipulate the undo headers in such a way that it will always restore the extmarks properly, and not undo previous changes. Might be better to make a snapshot of all the extmarks and restore them along with the original text or something.